File: docs/chapter1/section3/src/4-2_fizzbuzz.go
package main

import (
	"fmt"
	"net/http"
	"strconv"

	"github.com/labstack/echo/v4"
)

func main() {
	e := echo.New()

	// 「/fizzbuzz」というエンドポイントを設定する
	e.GET("/fizzbuzz", fizzBuzzHandler)

	e.Logger.Fatal(e.Start(":8080"))
}

func fizzBuzzHandler(c echo.Context) error {
	// クエリパラメーター "count" の値をcountStrに格納
	countStr := c.QueryParam("count")
	// もしcountStrが空っぽ("")だったら、
	if countStr == "" {
		// countStrを30に設定
		countStr = "30"
	}

	// strconv.Atoi()でstringをintに変換する(分からなかったらググりましょう！)
	count, err := strconv.Atoi(countStr)
	// ↑でエラーが起きたら、"count" が整数として解釈できないということなので、
	if err != nil {
		// ステータスコード 400 Bad Request を返す
		return c.String(http.StatusBadRequest, "Bad Request\n")
	}

	// fizzBuzzの処理を行う
	fizzBuzzStr := fizzBuzz(count)

	// ステータスコード 200 Ok とfizzBuzzの結果を返す
	return c.String(http.StatusOK, fizzBuzzStr)
}

// fizzBuzzの処理。これは競プロ
func fizzBuzz(n int) string {
	result := ""
	for i := 1; i <= n; i++ {
		fizz := i%3 == 0
		buzz := i%5 == 0

		if fizz && buzz {
			result += "FizzBuzz\n"
		} else if fizz {
			result += "Fizz\n"
		} else if buzz {
			result += "Buzz\n"
		} else {
			result += fmt.Sprintf("%d\n", i)
		}
	}

	return result
}

File: docs/chapter1/section3/src/4-2_fizzbuzz.rs
use axum::{extract::Query, http::StatusCode, routing::get, Router};
#[tokio::main]
async fn main() {
    // 「/ping」というエンドポイントを設定する
    let app = Router::new().route("/fizzbuzz", get(fizzbuzz_handler));

    // ポート8080でリスナーを作成する
    let listener = tokio::net::TcpListener::bind("127.0.0.1:8080")
        .await
        .unwrap();

    println!("listening on {}", listener.local_addr().unwrap());

    // サーバーを起動する
    axum::serve(listener, app).await.unwrap();
}

// クエリパラメータを受け取るための構造体を定義
#[derive(serde::Deserialize)]
struct FizzBuzzQuery {
    count: Option<String>,
}

async fn fizzbuzz_handler(Query(query): Query<FizzBuzzQuery>) -> (StatusCode, String) {
    // クエリパラメータが指定されていない場合はデフォルト値を使用する
    let mut n: i32 = 30;
    // クエリパラメータが指定されている場合はその値を調べる
    if let Some(count) = query.count {
        let count = count.parse();
        match count {
            // 数値に変換できた場合はその値を使用する
            Ok(count) => n = count,
            // ステータスコード 400 Bad Request を返す
            Err(_) => return (StatusCode::BAD_REQUEST, String::from("Bad Request\n")),
        }
    }

    // FizzBuzzの処理をする
    let fizzbuzz_str = fizzbuzz(n);

    // ステータスコード 200 Ok とfizzBuzzの結果を返す
    (StatusCode::OK, fizzbuzz_str + "\n")
}

// fizzBuzzの処理
fn fizzbuzz(n: i32) -> String {
    (1..=n)
        .map(|i| match (i % 3, i % 5) {
            (0, 0) => "FizzBuzz".to_string(),
            (0, _) => "Fizz".to_string(),
            (_, 0) => "Buzz".to_string(),
            (_, _) => i.to_string(),
        })
        .collect::<Vec<String>>()
        .join("\n")
}
